### PR TITLE
perf: add in-memory caches to reduce DB and Valkey call frequency

### PR DESCRIPTION
**Pull request by karaktaka**
## Summary

Extends the existing `GuildConfigCache` pattern and adds TTL-based caches across bot and web to eliminate redundant DB round-trips on hot paths.

### Bot side

- **Reaction role mappings** — `GuildConfigCache` now holds the full `message_id → {emoji: role_id}` mapping. `_get_role_for_reaction()` hits the cache first; DB is only consulted on a true miss, and the result is backfilled. Admin commands (`_reactionrole_add/remove`) keep the cache in sync.
- **Leave message config** — Removed the redundant `_leave_guild_ids` set; `_leave_configs` dict (`guild_id → (channel_id, message)`) is the single source of truth. `on_member_remove` reads directly from the cache — no DB session opened for the happy path.
- **Crafting recipe queries** — `_cache_recipe_query` decorator wraps all 16 read-only `CraftingRecipeCache` classmethods with a 1-hour `TTLCache`. Cache is explicitly cleared after each recipe sync in `blizzard.py`.
- **Autocomplete handlers** — Shared `TTLCache(maxsize=500, ttl=30)` via `cached_autocomplete()` helper applied to all 6 autocomplete methods across `roles`, `tagging`, `application`, and `music` modules.

### Web side

- **Premium user checks** — `require_premium` and `/me` use a 5-minute TTL cache for the premium user ID set; cleared explicitly on grant/revoke in `operator.py`.
- **Bot guild IDs** — `/me` caches `BotGuild.get_ids()` with a 2-minute TTL instead of a full table scan on every auth check.
- **Guild language** — The single `strings.get_guild_language()` call in `guilds.py` is wrapped in a 5-minute TTL cache.

### Test isolation

Added autouse fixtures to clear all module-level caches between tests (bot: `tests/conftest.py`, models: `tests/models/conftest.py`, web: `tests/web/conftest.py`).

## Test plan

- [ ] `uv run python -m pytest` — all 1111 tests pass
- [ ] `uv run ruff check` — no lint errors
- [ ] Reaction roles: add a reaction role, react to message, verify role assignment
- [ ] Leave messages: enable leave message, member leaves, verify message sent
- [ ] Crafting orders: open wizard, click through selections, verify no regressions
- [ ] Dashboard: log in as premium user, verify guild access still works

Closes #338

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * /playlist now guild-only; in-process caches added for recipes, autocompletes, premium status, guild language, reaction-roles, modroles, and leave-message configs.

* **Performance Improvements**
  * Fewer database queries via cached autocompletes and guild-level caches with targeted invalidation.

* **Reliability / Bug Fixes**
  * Better DB-error handling, stricter logging, conservative cache rewarm/eviction, and hidden replies for moderation leave-message commands.

* **Tests**
  * New and expanded tests plus autouse fixtures to validate cache behavior, invalidation, and edge cases.

### CODE REVIEW

File: NerdyPy/utils/cache.py [hunk 1]
```diff
@@ -160,7 +254,42 @@ def evict_guild(self, guild_id: int) -> None:
...
+    return result
+
+
+def build_name_choices(names: list[str], current: str) -> list[app_commands.Choice[str]]:
+    """Filter a list of names by the current autocomplete input and return up to 25 choices.
+
+    Companion to ``cached_autocomplete`` for the common case where the cache stores
+    plain name strings and Discord choices have name == value.
+    """
+    return [app_commands.Choice(name=n[:100], value=n) for n in names if current.lower() in n.lower()][:25]
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/cache.py]: Fixed in 7f37ad9 — both `name` and `value` are now truncated to 100 chars.

File: NerdyPy/utils/cache.py [hunk 2]
```diff
@@ -160,7 +254,42 @@ def evict_guild(self, guild_id: int) -> None:
...
+        fetcher: Zero-argument callable that opens a DB session and returns a list.
+
+    Returns:
+        The cached or freshly fetched list.
+    """
+    if key in _autocomplete_cache:
+        return _autocomplete_cache[key]
+    result = fetcher()
+    _autocomplete_cache[key] = result
+    return result
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/cache.py]: Fixed in ffee3ce — `cached_autocomplete` is now `async` and wraps the fetcher in `asyncio.to_thread()`. All call sites updated to `await` the call. Test engine updated to use `StaticPool` so worker threads share the in-memory SQLite connection.

File: NerdyPy/utils/cache.py [hunk 3]
```diff
@@ -83,39 +118,62 @@ def is_reaction_role_message(self, message_id: int) -> bool:
...
         return message_id in self._rr_message_ids
 
+    def get_reaction_role(self, message_id: int, emoji: str) -> int | None:
+        """Return the role ID for the given message + emoji, or None if not mapped.
+
+        Returns None before warm-up completes (callers must fall back to DB).
+        """
+        if not self._rr_warmed:
+            return None
+        return self._rr_mappings.get(message_id, {}).get(emoji)
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/cache.py]: Fixed: `REACTION_ROLE_CACHE_MISS = object()` sentinel added in `cache.py`. `get_reaction_role()` now returns the sentinel when not yet warmed, `None` when warmed-but-unmapped, and an `int` role ID when mapped. `_get_role_for_reaction()` in `roles.py` distinguishes all three cases — sentinel falls back to DB (and backfills), `None` short-circuits without a DB hit.

File: NerdyPy/utils/valkey.py [hunk 4]
```diff
@@ -262,6 +262,26 @@ def _call_api():
...
+            await to_thread(bot.guild_cache.reload_leave_config, guild_id, bot.SESSION)
+        except Exception:
+            bot.log.exception("invalidate_leave_config: reload failed for guild_id=%d", guild_id)
+            return {"ok": False, "error": "cache reload failed — see bot logs"}
+        return {"ok": True}
+    elif command == "invalidate_modrole":
+        guild_id = int(payload.get("guild_id", 0))
+        if not guild_id:
+            bot.log.warning("invalidate_modrole: received invalid guild_id=%r", payload.get("guild_id"))
+            return {"ok": False, "error": "invalid guild_id"}
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/valkey.py:310]: Fixed in the latest push: each of the three new command handlers now wraps the `int()` call in a `try/except (TypeError, ValueError)` with an `or 0` guard, so `None` and non-numeric strings both land in the dedicated `{ok: False, error: 'invalid guild_id'}` path instead of raising.

File: web/routes/guilds.py [hunk 5]
```diff
@@ -56,6 +57,32 @@
...
+    try:
+        return _guild_lang_cache[guild_id]
+    except KeyError:
+        pass
+    from models.admin import GuildLanguageConfig
+
+    try:
+        config = GuildLanguageConfig.get(guild_id, session)
+    except SQLAlchemyError:
+        _log.exception("_get_guild_language_cached: DB read failed for guild_id=%d", guild_id)
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply web/routes/guilds.py:79]: False positive —  is typed  as a FastAPI path parameter. FastAPI validates and coerces path parameters before the handler runs; a non-integer value raises HTTP 422 before this function is ever reached. An  cannot carry log-injection payloads, and `%d` formats it as a plain decimal integer regardless. No meaningful injection surface exists here.

File: NerdyPy/modules/application.py [hunk 6]
```diff
@@ -78,39 +74,54 @@ async def cog_load(self):
...
     ) -> list[app_commands.Choice[str]]:
-        with self.bot.session_scope() as session:
-            return _filter_choices(ApplicationTemplate.get_guild_templates(interaction.guild.id, session), current)
+        guild_id = interaction.guild.id
+
+        def _fetch():
+            with self.bot.session_scope() as session:
+                return [t.Name for t in ApplicationTemplate.get_guild_templates(guild_id, session)]
+
+        return build_name_choices(await cached_autocomplete(("app_guild_templates", guild_id), _fetch), current)
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/modules/application.py:124]: Out of scope for this PR — tracked in #403. The fix requires changing the autocomplete closures to return (Id, Name) tuples and updating all downstream resolver code to look up by primary key. In practice, Discord bot form names rarely hit 100 chars, but the design gap is real.

File: NerdyPy/utils/cache.py [hunk 7]
```diff
@@ -124,43 +221,194 @@ def is_leave_message_guild(self, guild_id: int) -> bool:
...
+            _log.exception("get_leave_config: DB read failed for guild_id=%d — returning None", guild_id)
+            return None
+
+        self._leave_configs[guild_id] = config
+        self._leave_evicted.discard(guild_id)
+        return config
+
+    def set_leave_config(self, guild_id: int, channel_id: int, message: str | None) -> None:
+        """Upsert the leave message config for a guild."""
+        self._leave_configs[guild_id] = (channel_id, message)
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/cache.py]: Fixed — added self._leave_evicted.discard(guild_id) to set_leave_config().

File: NerdyPy/utils/valkey.py [hunk 8]
```diff
@@ -262,6 +284,38 @@ def _call_api():
...
+        guild_id = _parse_guild_id(payload)
+        if not guild_id:
+            bot.log.warning("invalidate_leave_config: received invalid guild_id=%r", payload.get("guild_id"))
+            return {"ok": False, "error": "invalid guild_id"}
+        try:
+            await to_thread(bot.guild_cache.reload_leave_config, guild_id, bot.SESSION)
+        except Exception:
+            bot.log.exception("invalidate_leave_config: reload failed for guild_id=%d", guild_id)
+            return {"ok": False, "error": "cache reload failed — see bot logs"}
+        return {"ok": True}
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply NerdyPy/utils/valkey.py:305]: Out of scope for this PR — tracked in #404. In CPython, the GIL makes the individual dict/set operations in reload_leave_config thread-safe, so this is a design concern rather than an active bug. The proper fix (split DB-fetch from cache-mutation) is a separate refactor.

File: web/cache.py [hunk 9]
```diff
@@ -128,6 +128,17 @@ async def send_bot_command(self, command: str, payload: dict, timeout: float = 3
...
+    def notify_bot(self, command: str, payload: dict) -> None:
+        """Publish a fire-and-forget command to the bot (no reply expected).
+
+        Failures are logged as errors — callers must not rely on delivery.
+        """
+        message = json.dumps({"command": command, **payload})
+        try:
+            self._client.publish(self._key("cmd"), message)
+        except Exception as exc:
+            _log.error("notify_bot: failed to publish %r: %s", command, exc, exc_info=True)
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply web/cache.py:140]: Not fixable as described. Re-raising from notify_bot would break the HTTP 200 contract for callers where the DB write succeeded — the cache invalidation is best-effort by design (the docstring says so). The underlying concern is valid: bot caches without TTL stay stale indefinitely if Valkey is unreachable. The correct fix is TTL fallback on the bot caches (or periodic revalidation), which is a larger design change. Leaving this open for a follow-up.

File: web/routes/guilds.py [hunk 10]
```diff
@@ -116,7 +140,11 @@ async def set_language(
         session.add(cfg)
     else:
         cfg.Language = body.language
-    return LanguageConfig(guild_id=str(guild_id), language=cfg.Language)
+    lang = cfg.Language
+    session.commit()
+    _guild_lang_cache[guild_id] = lang
+    vk.notify_bot("set_guild_language", {"guild_id": str(guild_id), "language": lang})
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply web/routes/guilds.py]: Not addressed — by design.  is the correct call here: it logs publish failures at `error` level (operators see them), and deliberately does not block the HTTP response on Valkey availability. The caches it writes (language, modrole, leave config) all have TTL or on-demand reload paths, so a transient Valkey outage causes at most one cache cycle of stale reads, not indefinite staleness. Blocking the route on `send_bot_command` would make language/modrole updates unavailable whenever the bot is offline or restarting — a much worse trade-off.

File: tests/modules/test_roles_reactionrole_events.py [hunk 11]
```diff
@@ -77,6 +77,85 @@ def _make_payload(message_id, guild_id, emoji_name, member=None, user_id=999):
...
+    ):
+        """Warm cache with message_id mapped to 👍 only; reaction for ❤ returns None without a DB call.
+
+        Pins the warm-but-mismatched-emoji path: get_reaction_role returns None in-memory,
+        and the cold-miss DB fallback is never triggered.
+        """
+        cog.bot.guild_cache.warm_reaction_roles(cog.bot.SESSION)
+
+        # Patch session_scope so any DB access would raise and fail the test.
+        def _no_db():
```
[the remark this answers is not in the document]
Comment (karaktaka) [reply tests/modules/test_roles_reactionrole_events.py]: Fixed in b030ddf — removed the unused `_no_db` function and replaced it with `cog.bot.session_scope = MagicMock(side_effect=AssertionError(_NO_DB_MSG))` wired up in all three warm-cache tests.
